controllers/horizontal_runner_autoscaler_webhook.go

```diff
@@ -489,55 +493,158 @@ func (autoscaler *HorizontalRunnerAutoscalerGitHubWebhook) getScaleUpTarget(ctx
-		)
-		return nil, nil
+	if enterprise != "" {
+		// Search for enterprise runner HRAs in default runner group
+		if target, err := autoscaler.getJobScaleTarget(ctx, enterpriseKey(enterprise), labels); err != nil {
+			log.Error(err, "finding enterprise runner", "enterprise", enterprise)
+			return nil, err
+		} else if target != nil {
+			log.Info("scale up target is default enterprise runners", "enterprise", enterprise)
+			return target, nil
+		}
 	}
 
-	if target, err := autoscaler.getJobScaleTarget(ctx, enterpriseKey(enterprise), labels); err != nil {
-		log.Error(err, "finding enterprise runner", "enterprise", enterprise)
+	// At this point there were no default organization/enterprise runners available to use, try now
+	// searching in runner groups
+
+	// We need to get the potential runner groups first to avoid spending API queries needless. Once/if GitHub improves an
+	// API to find related/linked runner groups from a specific repository this logic could be removed
+	availableEnterpriseGroups, availableOrganizationGroups, err := autoscaler.getPotentialGroupsFromHRAs(ctx, enterprise, owner)
+	if err != nil {
+		log.Error(err, "finding potential organization runner groups from HRAs", "organization", owner)
 		return nil, err
-	} else if target != nil {
-		log.Info("scale up target is enterprise runners", "enterprise", enterprise)
-		return target, nil
-	} else {
+	}
+	if len(availableEnterpriseGroups) == 0 && len(availableOrganizationGroups) == 0 {
 		log.V(1).Info("no repository/organizational/enterprise runner found",
 			"repository", repositoryRunnerKey,
 			"organization", owner,
 			"enterprises", enterprise,
 		)
 	}
 
+	var enterpriseGroups []string
+	var organizationGroups []string
+	if autoscaler.GitHubClient != nil {
+		// Get available organization runner groups and enterprise runner groups for a repository
+		// These are the sum of runner groups with repository access = All repositories plus
+		// runner groups where owner/repo has access to
+		enterpriseGroups, organizationGroups, err = autoscaler.GitHubClient.GetRunnerGroupsFromRepository(ctx, owner, repositoryRunnerKey, availableEnterpriseGroups, availableOrganizationGroups)
+		log.V(1).Info("Searching in runner groups", "enterprise.groups", enterpriseGroups, "organization.groups", organizationGroups)
+		if err != nil {
+			log.Error(err, "Unable to find runner groups from repository", "organization", owner, "repository", repo)
+			return nil, nil
+		}
+	} else {
+		// For backwards compatibility if GitHub authentication is not configured, we assume all runner groups have
+		// visibility=all to honor the previous implementation, therefore any available enterprise/organization runner
+		// is a potential target for scaling
+		enterpriseGroups = availableEnterpriseGroups
+		organizationGroups = availableOrganizationGroups
+	}
+
+	for _, group := range organizationGroups {
+		if target, err := autoscaler.getJobScaleTarget(ctx, owner+"/"+group, labels); err != nil {
```

Comment:
I think this can break when there were both (1) a repository named `group` in the org and (2) runner group named `group` in the organization. If that's the case we might better use a dedicated key prefix for this case?

Reply:
yep noticed that same I'm preparing a new commit where I'm using for runner groups `owner+"/group/"+group`

Reply:
Added in 200518c45b9ce48872075536d2d0f5463c3c9d0e. I've also try to remove some duplicate code between `getScaleUpTarget` and `getJobScaleUpTargetForRepoOrOrg`

Reply:
Ideally, logic change and refactoring should be a separate commit for easy reviews but let me try!
Thanks anyway for your prompt response. I'll do a full review in a day or two!

Reply:
Awesome - thanks @mumoshu